src/api/libcellml/issue.h

```diff
@@ -0,0 +1,433 @@
+/*
+Copyright 2016 University of Auckland
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include "libcellml/exportdefinitions.h"
+#include "libcellml/referencerules.h"
+#include "libcellml/types.h"
+
+#include <string>
+
+namespace libcellml {
+
+/**
+ * @brief The Issue class
+ *
+ * Base class for issues used with logger derived classes.
+ */
+class LIBCELLML_EXPORT Issue
+{
+public:
+    virtual ~Issue(); /**< Destructor */
+    Issue(const Issue &rhs) = delete; /**< Copy constructor */
+    Issue(Issue &&rhs) noexcept = delete; /**< Move constructor */
+    Issue &operator=(Issue rhs) = delete; /**< Assignment operator */
+
+    /**
+     * @brief Create an @c Issue object.
+     *
+     * Factory method to create an @c Issue.  Can create a
+     * blank issue with::
+     *
+     *   IssuePtr issue = libcellml::Issue::create();
+     *
+     * or an issue with one of the following types as a parameter::
+     *
+     *   - libcellml::ComponentPtr
+     *   - libcellml::ImportSourcePtr
+     *   - libcellml::ModelPtr
+     *   - libcellml::ResetPtr
+     *   - libcellml::UnitsPtr
+     *   - libcellml::VariablePtr
+     *
+     * @return A smart pointer to an @c Issue object.
+     */
+    static IssuePtr create() noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const ComponentPtr &component) noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const ImportSourcePtr &importSource) noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const ModelPtr &model) noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const ResetPtr &reset) noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const UnitsPtr &units) noexcept;
+
+    /**
+     * @overload
+     */
+    static IssuePtr create(const VariablePtr &variable) noexcept;
+
+    /**
+     * @brief The issue Cause enum class.
+     *
+     * Enum to describe the cause of issue a given issue is.
+     */
+    enum class Cause
```

Comment:
shouldn't there be a constructor and get/set methods for all the different `Cause`s?

Reply:
I don't think so.  `Cause` replaces what was named `Kind`, so it's a straight swap from what was previously there ... I only changed the name as having `Level` and `Kind` was too confusing.